### PR TITLE
📍 Support training peft model with gradient checkpointing

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/test_sft_trainer.py [hunk 1]
```diff
@@ -799,6 +799,42 @@ def test_train_with_peft_config_gradient_checkpointing(self):
...
+        model = get_peft_model(model, LoraConfig())
+
+        # Get the dataset
+        dataset = load_dataset("trl-internal-testing/zen", "standard_language_modeling", split="train")
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            # Initialize the trainer
+            training_args = SFTConfig(output_dir=tmp_dir, gradient_checkpointing=True, report_to="none")
+
+            trainer = SFTTrainer(model=model_id, args=training_args, train_dataset=dataset)
```
Comment (qgallouedec) [inline tests/test_sft_trainer.py]:
```suggestion
            trainer = SFTTrainer(model=model, args=training_args, train_dataset=dataset)
```

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -401,7 +401,7 @@ def __init__(
                 model, processing_class = clone_chat_template(model, processing_class, args.chat_template_path)
 
         # PEFT configuration and model wrapping
-        if peft_config is not None:
+        if peft_config is not None or (is_peft_available() and isinstance(model, PeftModel)):
```
Comment (kashif) [inline trl/trainer/sft_trainer.py:404]: is this needed? as `_prepare_peft_model` checks for these two conditions and either returns an error or returns back the model?

**Review by kashif — APPROVED** [review]

File: tests/test_sft_trainer.py [hunk 3]
```diff
@@ -799,6 +799,42 @@ def test_train_with_peft_config_gradient_checkpointing(self):
...
+        model = get_peft_model(model, LoraConfig())
+
+        # Get the dataset
+        dataset = load_dataset("trl-internal-testing/zen", "standard_language_modeling", split="train")
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            # Initialize the trainer
+            training_args = SFTConfig(output_dir=tmp_dir, gradient_checkpointing=True, report_to="none")
+
+            trainer = SFTTrainer(model=model, args=training_args, train_dataset=dataset)
```
Comment (kashif) [inline tests/test_sft_trainer.py:817]:
```suggestion
            trainer = SFTTrainer(model=model, args=training_args, train_dataset=dataset)
            
            # Verify model is a PeftModel
            self.assertIsInstance(trainer.model, PeftModel)
```
